app/jobs/idd2wdl.py

```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python3
+
+import os
+import json
+import argparse
+import collections
+
+parser = argparse.ArgumentParser("idd2wdl", description="Convert an idseq-dag DAG to a WDL workflow")
+parser.add_argument("dag")
+parser.add_argument("--name")
+parser.add_argument("--pipeline-version")
+args = parser.parse_args()
+workflow_name = args.name or os.path.basename(args.dag).replace(".json", "")
+
+with open(args.dag) as fh:
+    dag = json.load(fh)
+
+print("version 1.0")
+
+input_names = collections.OrderedDict()
+targets_to_steps, step_class_count = dict(), collections.defaultdict(int)
+for step in dag["steps"]:
+    targets_to_steps[step["out"]] = step
+    step_class_count[step["class"]] += 1
+
+
+def file_path_to_name(f):
+    return f.replace(".", "_").replace("/", "_").replace("-", "_")
+
+
+def task_name(step):
+    if step_class_count[step["class"]] < 2:
+        name = step["class"]
+    else:
+        name = step["class"] + "_" + step["out"]
+    if name.startswith("PipelineStep"):
+        name = name[len("PipelineStep"):]
+    return name
+
+
+def target_filename(target, index):
+    if target == "fastqs":
+        return target + "_" + str(index)
+    else:
+        return target + "_" + file_path_to_name(dag["targets"][target][index])
+
+
+for step in dag["steps"]:
+    idd_step_input, wdl_step_input, input_files, input_files_local = [], [], [], []
+    for target in step["in"]:
+        idd_step_input.append(dag["targets"][target])
+        input_files.append([])
+        input_files_local.append([])
+        for i, file_target in enumerate(dag["targets"][target]):
+            if target in targets_to_steps:
+                input_name = file_path_to_name(file_target)
+            else:
+                input_name = target_filename(target, i)
+                input_names[input_name] = 1
+            wdl_step_input.append("    File " + input_name)
+            input_files[-1].append(None)
+            input_files_local[-1].append("~{" + input_name + "}")
+    wdl_step_input = "\n".join(wdl_step_input)
+    input_files_local = json.dumps(input_files_local)
+    idd_step_output = dag["targets"][step["out"]]
+    wdl_step_output = "\n".join('    File {} = "{}"'.format(file_path_to_name(name), name) for name in idd_step_output)
+    s3_wd_uri = os.path.join(dag["output_dir_s3"], "main/sfn-1/wdl-1/dag-" + str(args.pipeline_version))
+    nha_cluster_ssh_key_uri = ""
+    if "environment" in step["additional_attributes"]:
+        nha_cluster_ssh_key_uri = "s3://idseq-secrets/idseq-{}.pem".format(step["additional_attributes"]["environment"])
+
+    print("""
+task {task_name} {{
+  runtime {{
+    docker: "{AWS_ACCOUNT_ID}.dkr.ecr.us-west-2.amazonaws.com/idseq-workflows:{STAGE}"
+  }}
+  input {{
+{wdl_step_input}
+  }}
+  command<<<
+  python3 <<CODE
+  import os, sys, json, contextlib, importlib, threading, logging, subprocess, traceback
```

Comment:
I have a suggestion for this code: could we create a static class in idseq_dag with all this logic below, and instead of adding this huge block here, just a static method sending a JSON entry. That would not only make the code more readable, but also would prevent potential command injection.

Ex:
- Create a `idseq-dag/wdl_task_wrapper.py`
```python3
def wdl_execute_task(inputJson):
   # move all the logic to this block
```

And here in the wdl template, just use:
```python3
task {task_name} {{
    ...
    command<<<
       python3 <<CODE
            from idseq_dag.wdl_task_wrapper import wdl_wrap_task
            wdl_execute_task({input_json})
    ...
    }}""".format(
        JSON.dumps({
             task_name=task_name(step),
             ...
        })
```

Reply:
For me, just depends how long this file is going to be here. @kislyuk:  wdyt ?

Reply:
The transition plan after we ship the initial version of this code is to stop using jbuilder templates and start using static WDL files representing each of the stages. In other words, this code (idd2wdl.py) will be removed from idseq-web as soon as we can get rid of the jbuilder templates. Any parameters that parameterize the jbuilder templates will be moved into the static WDL interface.

The suggestion to keep calling idseq-dag to translate the dynamically generated DAG (using facilities internal to idseq-dag) breaks this interface migration, and takes us further away from a clean separation between the front-end and the pipeline. So I think we should stick with the transition plan as we have defined it so far.

Reply:
Actually, reading further into David's proposal, I can see that it concerns just the python code in the WDL command block. I initially misread it as concerning the whole idd2wdl.py layer.

I think it would be reasonable to put this code in a `idseq-dag/wdl_task_wrapper.py` file instead of autogenerating it here. I can go ahead and implement that. I'll post here when I get a PR merged in idseq-dag, release it, and rebuild a container image for the sfn-wdl pipeline with the new idseq-dag version.

Reply:
Because the task of moving code to idseq-dag will necessitate additional testing and validation, I would like to postpone it and break it out into a separate sub-task, without blocking this PR on it.

Reply:
As an aside, I consider the command injection concern to be unwarranted, since the inputs are extracted from DAG JSON.

Reply:
Filed https://jira.czi.team/browse/IDSEQ-2390 to track the additional task.